lib/dradis/plugins/mapping.rb

```diff
@@ -0,0 +1,60 @@
+
+# When you call provides :upload in your Engine, this module gets included.
+module Dradis::Plugins::Mapping
+  extend ActiveSupport::Concern
+
+  included do
+    module_parent.extend ClassMethods
+  end
+
+  module ClassMethods
+
+    def component
+      meta[:name].to_s
+    end
+
+    def field_names(source:, destination: nil, field_type: 'destination')
+      mappings = Mapping.includes(:mapping_fields).where(component: component)
+      mappings = mappings.where(source: source) if source
+      mappings = mappings.where(destination: destination) if destination
+
+      fields = mappings.pluck("mapping_fields.#{field_type}_field").uniq
+
+      if fields.empty? && source
+        default_mapping(source).keys
+      else
+        fields
+      end
+    end
+
+    def default_mapping(source)
+      self::Mapping::DEFAULT_MAPPING[source.to_sym]
+    end
+
+    def mapping(source:, destination:)
+      Mapping.includes(:mapping_fields).find_by(
+        component: component,
+        source: source,
+        destination: destination
+      )
+    end
+
+    def mapping_fields(source:, destination:)
```

Comment:
Isn't this code very very similar to #field names? can we have the former call this and then manipulate the results?

Reply:
`field_names` pulls all field names from all mappings for the given arguments, whereas `mapping_fields` pulls the fields for one mapping.
I updated the `mapping` method to return a collection of all mappings matching the given source and/or destination, and `mapping_fields` will now return a collection of mapping fields for all mappings in the collection
This way, they can be used for a single mapping (applying mappings) or multiple matching mappings (rules engine field list)